paddle/framework/op_registry.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include <unordered_map>
+#include "paddle/framework/attr_checker.h"
+#include "paddle/framework/enforce.h"
+#include "paddle/framework/op_base.h"
+#include "paddle/framework/op_proto.pb.h"
+
+namespace paddle {
+namespace framework {
+
+// this class not only make proto but also init attribute checkers.
+class OpProtoAndCheckerMaker {
+ public:
+  OpProtoAndCheckerMaker(OpProto* proto, OpAttrChecker* op_checker)
+      : proto_(proto), op_checker_(op_checker) {}
+
+ protected:
+  void AddInput(const std::string& name, const std::string& comment) {
+    // add input to proto_
+  }
+  void AddOutput(const std::string& name, const std::string& comment) {
+    // add output to proto_
+  }
+  template <typename T>
+  TypedAttrChecker<T>& AddAttr(const std::string& name,
+                               const std::string& comment) {
+    // add attribute to proto_
+    return op_checker_->AddAttrChecker<T>(name);
+  }
+  void AddType(const std::string& op_type) {
+    // add type to proto_
+  }
+  void AddComment(const std::string& comment) {
+    // add comment to proto_
+  }
+
+  OpProto* proto_;
+  OpAttrChecker* op_checker_;
+};
+
+class OpRegistry {
+  typedef std::function<OperatorBase*(OpDesc& op_desc)> OpCreator;
```

Comment:
In `TypedAttrChecker::operator()`, we check attributes and **complete default value**, so the `OpDesc` will be mutated in inner checkers.